jax/_src/lru_cache.py

```diff
@@ -0,0 +1,127 @@
+# Copyright 2022 The JAX Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+import os
+import tensorflow as tf
+from typing import Optional
+
+from filelock import FileLock
+
+from jax._src.compilation_cache_interface import CacheInterface
+from jax._src.gfile_cache import GFileCacheImpl
+
+
+logger = logging.getLogger(__name__)
+
+
+def is_path_gs(path: str) -> bool:
+    return path.startswith("gs://")
+
+
+def get_size(path: str) -> int:
+    return tf.io.gfile.stat(path).length
+
+
+def get_mtime(path: str) -> int:
+    return tf.io.gfile.stat(path).mtime_nsec
+
+
+def update_mtime(path: str) -> None:
+    if not is_path_gs(path):
+        os.utime(path)  # set mtime (and also atime) to current time
+    else:
+        tmp_path = f"{path}.tmp"
+        tf.io.gfile.rename(path, tmp_path)
+        tf.io.gfile.rename(tmp_path, path)
+
+
+class LRUCacheImpl:
+
+    def __init__(self, cache_dir: str, max_cache_size: int, timeout=10):
+        self.cache_dir = cache_dir
+        self.max_cache_size = max_cache_size
+        os.makedirs(cache_dir, exist_ok=True)
+        self.timeout = timeout
+
+        self.lock_file = os.path.join(self.cache_dir, ".lockfile")
+        self.lock = FileLock(self.lock_file)
+
+    def get(self, key: str) -> Optional[bytes]:
+        with self.lock.acquire(timeout=self.timeout):
+            file_path = os.path.join(self.cache_dir, key)
+
+            if not os.path.exists(file_path):
+                logger.debug("Cache miss")
+                return None
+
+            logger.debug("Cache hit")
+            update_mtime(file_path)
+            with open(file_path, "rb") as f:
+                return f.read()
+
+    def put(self, key: str, val: bytes) -> None:
+        with self.lock.acquire(timeout=self.timeout):
+            file_path = os.path.join(self.cache_dir, key)
+            if os.path.exists(file_path):
+                return
+
+            self._evict_if_needed()
+            with open(file_path, "wb") as f:
+                f.write(val)
+            update_mtime(file_path)
+
+    def _evict_if_needed(self) -> None:
+        if self.max_cache_size == -1:
+            return  # max_cache_size == -1: no limit on cache size
+
+        files = os.listdir(self.cache_dir)
```

Comment:
Am I right that we assume that some other process can modify `cache_dir` concurrently? If not, it could make sense to keep track of the number of cache entries in memory.

Reply:
Yes, we need to consider that some other process can modify `cache_dir` concurrently.